### PR TITLE
decode encoded passwords from properties file

### LINKED ISSUE #21845: featureUtility - Not decoding repository passwords when executing
**Describe the bug**  
When executing the featureUtility on `kernel-slim` images when ever we set overrides for the mavenCentralMirror or a secondary repository and we use the securityUtility to {xors}, {hash} or {aes} encode the passwords, authentication is failing to those end servers with http 401 forbidden. Replace it with the actual passwords in `plain text` and it works again.

**Steps to Reproduce**  
- Configure a on-prem or personal extra mirror  or override mavenCentralMirror
- Use securityUtils to {xors}, {hash} or {aes} encode your password configuration
- Execute featureUtils install (server or any normal install of features)

**Expected behavior**  
Expected that securityUtility encoded/hashed passwords suffice in the MavenRepository passwords.

**Diagnostic information:**  
 - OpenLiberty Version: [e.g. 21.0.0.12 - 22.0.0.6]
 - Affected feature(s) [Features not specific]
 - Java Version: OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
 - server.xml configuration (not needed to test issue)

**Additional context**  
The OpenLiberty documentation listed here: https://openliberty.io/docs/latest/reference/command/featureUtility-modifications.html
Shows that the repo2 password is {xor} hashed, thus why we think this is a bug.

``` 
[25/7/22, 20:23:02:272 AEST] Testing connection for repository: http://myreposerver/repository/maven-central/
[25/7/22, 20:23:02:403 AEST] Response code: 401
[25/7/22, 20:23:02:403 AEST] Connecting to the following repository: http://myreposerver/repository/maven-central/
[25/7/22, 20:23:02:525 AEST] CWWKF1392E: Incorrect credentials were provided for the http://myreposerver/repository/maven-central/ repository.
[25/7/22, 20:23:02:526 AEST] Could not download this json with maven coordinate: io.openliberty.features:features:21.0.0.12
[25/7/22, 20:23:02:526 AEST] Downloaded the following json files from remote: []
[25/7/22, 20:23:02:527 AEST] action.exception.stacktrace: null
[25/7/22, 20:23:02:527 AEST] CWWKF1409E: Unable to find the following feature JSON files locally or on the configured Maven repository: [io.openliberty.features:features:21.0.0.12].
com.ibm.ws.install.InstallException: CWWKF1409E: Unable to find the following feature JSON files locally or on the configured Maven repository: [io.openliberty.features:features:21.0.0.12].
        at com.ibm.ws.install.featureUtility.FeatureUtility.getJsonFiles(FeatureUtility.java:656)
        at com.ibm.ws.install.featureUtility.FeatureUtility.<init>(FeatureUtility.java:164)
        at com.ibm.ws.install.featureUtility.FeatureUtility.<init>(FeatureUtility.java:58)
        at com.ibm.ws.install.featureUtility.FeatureUtility$FeatureUtilityBuilder.build(FeatureUtility.java:792)
        at com.ibm.ws.install.featureUtility.cli.InstallServerAction.install(InstallServerAction.java:259)
        at com.ibm.ws.install.featureUtility.cli.InstallServerAction.execute(InstallServerAction.java:283)
        at com.ibm.ws.install.featureUtility.cli.InstallServerAction.handleTask(InstallServerAction.java:78)
        at com.ibm.ws.install.featureUtility.cli.FeatureAction.handleTask(FeatureAction.java:100)
        at com.ibm.ws.install.featureUtility.FeatureUtilityExecutor.main(FeatureUtilityExecutor.java:58)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.ibm.ws.kernel.boot.cmdline.UtilityMain.internal_main(UtilityMain.java:173)
        at com.ibm.ws.kernel.boot.cmdline.UtilityMain.main(UtilityMain.java:53)
        at com.ibm.ws.kernel.boot.cmdline.Main.main(Main.java:52)
```

I believe this can be resolved in the `ArtifactDownloader.java` class where it pulls out the password and use the `PasswordUtils.decode()` function before it - I may attempt to do a PR for this.

### PR DESCRIPTION
**Pull request by jjiwooLim**
Fixes #21845

### CODE REVIEW

**Comment by jjiwooLim** [comment]
#build

File: dev/com.ibm.ws.install.featureUtility/src/com/ibm/ws/install/featureUtility/props/FeatureUtilityProperties.java [hunk 1]
```diff
@@ -153,8 +162,26 @@ private static boolean parseProperties(Properties properties) {
...
+			    try {
+				decodedPwd = PasswordUtil.decode(value.toString());
+			    } catch (InvalidPasswordDecodingException ipde) {
+				logger.log(Level.FINE,
+					Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_TOOL_PWD_NOT_ENCRYPTED")
+						+ InstallUtils.NEWLINE);
+			    } catch (UnsupportedCryptoAlgorithmException ucae) {
+				throw new InstallException(
+					Messages.INSTALL_KERNEL_MESSAGES
+						.getLogMessage("ERROR_TOOL_PROXY_PWD_CRYPTO_UNSUPPORTED"),
```
Comment (veenarm) [inline dev/com.ibm.ws.install.featureUtility/src/com/ibm/ws/install/featureUtility/props/FeatureUtilityProperties.java]: Should this log message match the one above on line 173?

File: dev/com.ibm.ws.install.featureUtility/src/com/ibm/ws/install/featureUtility/props/FeatureUtilityProperties.java [hunk 2]
```diff
@@ -153,8 +162,26 @@ private static boolean parseProperties(Properties properties) {
...
+					ucae, InstallException.RUNTIME_EXCEPTION);
+			    }
+
```
Comment (veenarm) [inline dev/com.ibm.ws.install.featureUtility/src/com/ibm/ws/install/featureUtility/props/FeatureUtilityProperties.java]: Missing 
```suggestion
         repoMap.get(repoName).put(option, decodedPwd);
```

File: dev/com.ibm.ws.install.featureUtility/src/com/ibm/ws/install/featureUtility/props/FeatureUtilityProperties.java [hunk 3]
```diff
@@ -153,8 +162,26 @@ private static boolean parseProperties(Properties properties) {
...
                     String option = split[split.length - 1]; // incase there are periods in the key, cant use [1]
+
                     if(repoMap.containsKey(repoName)){
-                        repoMap.get(repoName).put(option, value.toString());
+			if (option.compareToIgnoreCase("password") == 0) {
+			    String decodedPwd = value.toString();
+			    try {
+				decodedPwd = PasswordUtil.decode(value.toString());
+			    } catch (InvalidPasswordDecodingException ipde) {
+				logger.log(Level.FINE,
```
Comment (veenarm) [inline dev/com.ibm.ws.install.featureUtility/src/com/ibm/ws/install/featureUtility/props/FeatureUtilityProperties.java]: Invalid password should follow same pattern as other places? to set decodedPwd back to the plain text from `value.toString()`

**Review by veenarm — CHANGES_REQUESTED** [review]
Currently doesn't look like the current PR will fix the issue as it's no longer put in the map.

**Comment by jjiwooLim** [comment]
#build

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java [hunk 4]
```diff
@@ -297,14 +301,26 @@ private void configureProxyAuthentication() {
         }
     }
 
-    private void configureAuthentication(final MavenRepository repository) {
+    private void configureAuthentication(final MavenRepository repository) throws InstallException {
```
Comment (jacobwdv) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java:296]: have you tested the scenario where the InstallException is thrown? I'm worried about throwing exceptions in methods that didn't throw them beforehand.

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java [hunk 5]
```diff
@@ -297,14 +301,26 @@ private void configureProxyAuthentication() {
...
 
         if (repository.getUserId() != null && repository.getPassword() != null &&
             envMap.get("https.proxyUser") == null && envMap.get("http.proxyUser") == null) {
+            String decodedPwd = repository.getPassword();
+            try {
+                decodedPwd = PasswordUtil.decode(repository.getPassword());
+            } catch (InvalidPasswordDecodingException ipde) {
+                decodedPwd = repository.getPassword();
```
Comment (jacobwdv) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java]: I think it's too assumptive that if we get InvalidPasswordDecodingException that it's just something we can ignore and assume that it's an unencoded password. From the API: 
```
If the encoded_string is null or invalid. Or the decoded_string is null.
```

Also from the API: 
```
Decode the provided string. The string should consist of the algorithm to be used for decoding and encoded string. For example, {xor}CDo9Hgw=. Use this method unless the encryption key needs to be specified for the AES encryption. 

 An empty algorithm "{}" is treated as not encoded. However, a missing algorithm will trigger the InvalidPasswordDecodingException.
```
I think we're assuming if we see this exception that the user provided a plain text password without the `{}`. But we instead detect that case and add them before hand easily: 
```
String encodedPassword = repository.getPassword();
        if (!encodedPassword.startsWith("{")) {
            encodedPassword = "{}" + encodedPassword;
        }
```

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java [hunk 6]
```diff
@@ -297,14 +301,26 @@ private void configureProxyAuthentication() {
...
+                logger.log(Level.FINE,
+                           Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_TOOL_PWD_NOT_ENCRYPTED")
+                                       + InstallUtils.NEWLINE);
+            } catch (UnsupportedCryptoAlgorithmException ucae) {
+                throw new InstallException(Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_TOOL_PWD_CRYPTO_UNSUPPORTED"));
+            }
+            final String pwd = decodedPwd;
```
Comment (jacobwdv) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java]: I'm not a fan of saving passwords in plain text variables. That might be a standard higher than we adhere to on open liberty but it's generally a safe practice to be scared of doing such a thing.

**Review by jacobwdv — CHANGES_REQUESTED** [review]
I think we need to reconsider being passive about `InvalidPasswordDecodingException`. We will likely break something down the chain if we pass it along 100% of the time. Instead I was thinking of a change like the one below. I have commented out creating the PasswordAuthentication object if we hit the InvalidPasswordDecodingException and instead throwing an InstallException but I will leave that up to you to consider.

 

```
    private void configureAuthentication(final MavenRepository repository) throws InstallException {

        if (repository.getUserId() != null && repository.getPassword() != null &&
            envMap.get("https.proxyUser") == null && envMap.get("http.proxyUser") == null) {
            final PasswordAuthentication auth = getPasswordAuthenticationFromRepository(repository);
            Authenticator.setDefault(new Authenticator() {
                @Override
                protected PasswordAuthentication getPasswordAuthentication() {
                    return auth;
                }
            });
        }
    }

    /**
     * @param repository - The repository to get the encoded password from
     * @return a PasswordAuthentication object with the username and decoded passwords saved within it
     * @throws InstallException if decoding the password fails due to an unsupported algorithm or invalid password.
     */
    private PasswordAuthentication getPasswordAuthenticationFromRepository(MavenRepository repository) throws InstallException {
        String encodedPassword = repository.getPassword();
        if (!encodedPassword.startsWith("{")) {
            encodedPassword = "{}" + encodedPassword;
        }
        PasswordAuthentication pa;
        try {
            pa = new PasswordAuthentication(repository.getUserId(), PasswordUtil.decode(encodedPassword).toCharArray());

        } catch (InvalidPasswordDecodingException ipde) {
            logger.log(Level.FINE, Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_TOOL_PWD_NOT_ENCRYPTED")
                                   + InstallUtils.NEWLINE,
                       ipde);
//            pa = new PasswordAuthentication(repository.getUserId(), repository.getPassword().toCharArray());
            throw new InstallException(Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_TOOL_PWD_NOT_ENCRYPTED"));

        } catch (UnsupportedCryptoAlgorithmException ucae) {
            logger.log(Level.SEVERE,
                       Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_TOOL_PWD_NOT_ENCRYPTED")
                                     + InstallUtils.NEWLINE,
                       ucae);
            throw new InstallException(Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_TOOL_PWD_CRYPTO_UNSUPPORTED"));
        }
        return pa;
    }

```

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java (hunk 5 above)
Comment (jacobwdv) [reply dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java]: Maybe we can't check for `{}` if it's AES encrypted, in which case we can't add the `{}`. Not sure, we would need to test it.

Comment (jacobwdv) [reply dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java]: Maybe we need to use `passwordDecode()` to support AES? either way both show that the algorithm should be provided in between `{` and `}` and unencoded passwords should still have `{}` in front to signal they are not encoded.

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java (hunk 4 above)
Comment (jjiwooLim) [reply dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java:296]: catching the exception in `testConnection(MavenRepository repository)` - line 341

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java [hunk 7]
```diff
@@ -79,9 +81,6 @@ private Future<String> submitDownloadRequest(String coords, String fileType, Str
 
     public Set<String> getMissingFeaturesFromRepo(List<String> mavenCoords, MavenRepository repository) throws InstallException {
         info(Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("STATE_CONTACTING_MAVEN_REPO"));
-        checkValidProxy();
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java:82]: Authentications will be configured in line 92 (when we call `testConnection(repository)`).

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java [hunk 8]
```diff
@@ -185,9 +184,6 @@ public void synthesizeAndDownload(String mavenCoords, String filetype, String dL
         checksumFormats[1] = "SHA1";
         checksumFormats[2] = "SHA256";
 
-        checkValidProxy();
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ArtifactDownloader.java:188]: Authentications will be configured in line 198 (when we call `testConnection(repository)`).

**Review by jacobwdv — APPROVED** [review]
Thank you, looks great!

**Comment by jjiwooLim** [comment]
#build

**Comment by jjiwooLim** [comment]
#libby-bot

**Pull request merged**